sklearn_numba_dpex/common/kernels.py

```diff
@@ -300,46 +270,40 @@ def make_sum_reduction_2d_kernel(
 
     .. [3] https://shreeraman-ak.medium.com/parallel-reduction-with-cuda-d0ae10c1ae2c
     """
-    if is_1d := not size1:
+    if is_1d := (len(shape) == 1):
         axis = 1
-        size1 = size0
-        size0 = 1
+        shape1 = shape[0]
+        shape0 = 1
+    else:
+        shape0, shape1 = shape
 
-    # NB: the shape of the work group is different in each of those two cases. Summing
-    # efficiently requires to adapt to very different IO patterns depending on the sum
-    # axis, which motivates the need for different kernels with different work group
-    # sizes for each cases. As a consequence, the shape of the intermediate results
-    # in the main driver loop, and the `global_size` (total number of work items fired
-    # per call) for each kernel call, are also different, and are variabilized with
-    # the lambda functions `get_result_shape` and `get_global_size`.
     if axis == 0:
-        work_group_size, kernels = _prepare_sum_reduction_2d_axis0(
-            size1,
+        work_group_shape, kernels, shape_update_fn = _prepare_sum_reduction_2d_axis0(
+            shape1,
             work_group_size,
             sub_group_size,
             fused_elementwise_func,
             dtype,
             device,
         )
-        get_result_shape = lambda result_sum_axis_size: (result_sum_axis_size, size1)
-        get_global_size = (
-            lambda result_sum_axis_size: result_sum_axis_size
-            * work_group_size
-            * math.ceil(size1 / sub_group_size)
-        )
     else:  # axis == 1
-        work_group_size, kernels = _prepare_sum_reduction_2d_axis1(
-            size0, work_group_size, fused_elementwise_func, dtype, device
-        )
-        get_result_shape = lambda result_sum_axis_size: (size0, result_sum_axis_size)
-        get_global_size = (
-            lambda result_sum_axis_size: result_sum_axis_size * work_group_size * size0
+        work_group_shape, kernels, shape_update_fn = _prepare_sum_reduction_2d_axis1(
+            shape0, work_group_size, fused_elementwise_func, dtype, device
         )
 
     # XXX: The kernels seem to work fine with work_group_size==1 on GPU but fail on CPU.
-    if work_group_size == 1:
+    if math.prod(work_group_shape) == 1:
         raise NotImplementedError("work_group_size==1 is not supported.")
```

Comment:
Should the message be changed here?

Reply:
In my mind `work_group_size=math.prod(work_group_shape)` so I think it's good.

There is this nuance because I have introduced the syntax `work_group_size=max` where we try to automatically use the maximum possible value for `work_group_size`. This change enable checking that `max` does not result in `work_group_size=1`.